release/rke2/images.go

```diff
@@ -0,0 +1,220 @@
+package rke2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log/slog"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	reg "github.com/rancher/ecm-distro-tools/registry"
+)
+
+type Architecture string
+
+const (
+	ArchLinuxAmd64   Architecture = "linux/amd64"
+	ArchLinuxArm64   Architecture = "linux/arm64"
+	ArchWindowsAmd64 Architecture = "windows/amd64"
+)
+
+type imageExpectations struct {
+	Reference         name.Reference
+	ExpectsLinuxAmd64 bool
+	ExpectsLinuxArm64 bool
+	ExpectsWindows    bool
+}
+
+type ImageStatus struct {
+	imageExpectations
+	OSSImage   reg.Image
+	PrimeImage reg.Image
+}
+
+type ReleaseInspector struct {
+	fs    fs.FS
+	oss   reg.Client
+	prime reg.Client
+	debug bool
+}
+
+func NewReleaseInspector(fs fs.FS, oss, prime reg.Client, debug bool) *ReleaseInspector {
+	return &ReleaseInspector{
+		fs:    fs,
+		oss:   oss,
+		prime: prime,
+		debug: debug,
+	}
+}
+
+func (r *ReleaseInspector) InspectRelease(ctx context.Context, version string) ([]ImageStatus, error) {
+	if !strings.Contains(version, "+rke2") {
+		return nil, errors.New("only RKE2 releases are currently supported")
+	}
+
+	archLists, err := r.getArchitectureLists()
+	if err != nil {
+		return nil, err
+	}
+	if r.debug {
+		slog.Debug("found architecture lists", "lists", mapKeys(archLists))
+	}
+
+	imageExpectations, err := r.processImageLists(archLists)
+	if err != nil {
+		return nil, err
+	}
+	if r.debug {
+		slog.Debug("processed images", "count", len(imageExpectations))
+	}
+
+	return r.checkImages(ctx, imageExpectations)
+}
+
+func mapKeys[K comparable, V any](m map[K]V) []K {
```

Comment:
First bit of parameterized types (not calling these generics since they're technically not) in the code base.

Reply:
I was playing around with them in this structured log, not here anymore. I've used some packages that are more like frameworks, where generics would be useful.